docs/adr/0007-azure-pipeline-security-model.md

```diff
@@ -0,0 +1,70 @@
+
+The challenge here is to balance the varied benefits of having a self-sufficient, repeatable pipeline with the need for adequate security controls to minimise the risk of the pipeline becoming the source of a security incident.
+
+Such a deployment pipeline needs to be able to perform the following actions:
+
+1. Create resources in the target subscription
+1. Query AAD for existing applications & service principals
+1. Perform ARM role assignments in the target subscription
+1. Create and manage new AAD applications & service principals
+
+### Create resources in the target subscription
+By default, the 'Service Connections' that Azure Pipelines use to authenticate with Azure only provide #1 of the above.
+
+### Perform ARM role assignments in the target subscription
+Typically this is achieved by granting the identity behind the service connection the 'owner' role in the target subscription.
+
+### Query AAD for existing applications & service principals
+The main pipeline is restructured to expect the required information to be provided as manually-maintained configuration.
+
+### Create and manage new AAD applications & service principals
+This is commonly handled in one of three ways:
+
+* an entirely seperate pipeline running under a different identity which does have the necessary permissions
+* running a subset of the main pipeline, under a different identity which does have the necessary permissions, such that only  the tasks that require the elevate privileges are run
+* an out-of-band process whereby an individual with the appropriate permissions executes the necessary steps (either manually or using a script)
+
+>NOTE: This functionality may have to run multiple times in the case of on-going application permission management.
+
+
+## Decision
+
+In order to ensure the principle of least privilege, whilst still enabling a fully-automated process the following steps are proposed.
+
+1. Each Marain environment has its own dedicated Azure AD identity that the pipeline executes under
+1. Each Azure subscription that hosts a Marain environment requires a custom AzureRM role, that grants:
+    * `MicrosoftAuthorization/RoleAssignment/*` (Read/Write/Delete)
+1. Each identity is granted the following permissions:
+    * Subscription scope:
+        * `Contributor`
+        * The custom role for the associated subscription
+    * Azure Active Directory Graph scope:
+        * `Directory.Read.All`
+        * `Application.ReadWrite.Owned`
+1. AzureAD administrators get approval oversight of the AzureAD Graph permissions, via the 'admin consent' mechanism
+1. An Azure Pipelines service connection is created for each of the above identities
+1. The owner of the Azure Pipelines service connection gets approval oversight of any pipelines wishing to use it
+1. By using YAML-based pipelines, any changes can be subjected to review via a Pull Request process
+
+
+## Consequences
```

Comment:
Are there any negative consequences?

Reply:
Yes. The ADO SP has role assignment privileges and contributer rights over the entire subscription in this solution. In cases where the subscription is not entirely dedicated to hosting this one Marain instance (and it's common for organizations to have one "production" sub instead of splitting it up into lots of little subs) this means the build pipeline has control over a lot of things it really has no business being able to control.

This is a direct consequence of granting this service principal subscription-scope privileges.

Reply:
Agreed - the only alternatives I could come up with were:
1. have the resource groups pre-created so the permission could be scoped to them
1. a process whereby these identities have their permissions managed on an on-going basis (rather than as a single up-front step, as this discussses)

Reply:
What's the problem with pre-creating resource groups?

Reply:
(and in any case, whether we go down that path or not, it would be good to address it as an alternative approach in this ADR, and explain why it's not currently our chosen option.)